orttraining/orttraining/python/training/ortmodule/_utils.py

```diff
@@ -97,3 +99,44 @@ def _create_iobinding(io_binding, inputs, model, device):
 
     for value_info in model.graph.output:
         io_binding.bind_output(value_info.name, device.type, device_id=get_device_index(device))
+
+
+class SessionConfig(NamedTuple):
+    session_options: onnxruntime.SessionOptions
+    providers: Sequence[Union[str, Tuple[str, Dict]]]
+    provider_options: Sequence[Dict]
+
+
+def get_session_config(device: torch.device, use_external_gpu_allocator: bool = True,
+                       is_rocm_pytorch: bool = False, loglevel: int = 2) -> SessionConfig:
+    """Creates and returns the session configuration to be used for the ExecutionAgent"""
+    providers = None
+    provider_options = None
+    if device.type == 'cuda':
+        # Configure the InferenceSessions to use the specific GPU on which the model is placed.
+        providers = (["ROCMExecutionProvider"] if is_rocm_pytorch else ["CUDAExecutionProvider"])
+        providers.append("CPUExecutionProvider")
+        if use_external_gpu_allocator and torch.cuda.is_available():
+            # CPP extension to get torch GPU allocator's alloc and free function addresses
+            from onnxruntime.training.ortmodule.torch_cpp_extensions import torch_gpu_allocator
+            torch_alloc = torch_gpu_allocator.gpu_caching_allocator_raw_alloc_address()
+            torch_free = torch_gpu_allocator.gpu_caching_allocator_raw_delete_address()
+            provider_options = [{"device_id": str(device.index),
+                                 "gpu_external_alloc": str(torch_alloc),
+                                 "gpu_external_free": str(torch_free)}, {}]
+        else:
+            provider_options = [{"device_id": str(device.index)}, {}]
+    elif device.type == 'cpu':
+        providers = ["CPUExecutionProvider"]
+        provider_options = [{}]
+
+    session_options = onnxruntime.SessionOptions()
+    session_options.enable_mem_pattern = False
+    session_options.enable_mem_reuse = False
+    session_options.use_deterministic_compute = False
+    # default to PRIORITY_BASED execution order
+    session_options.execution_order = onnxruntime.ExecutionOrder.PRIORITY_BASED
+    # 0:Verbose, 1:Info, 2:Warning. 3:Error, 4:Fatal. Default is 2.
+    session_options.log_severity_level = int(loglevel)
+
+    return SessionConfig(session_options, providers, provider_options)
```

Comment:
```suggestion
class SessionConfig(object):

    def __init__(self, device: torch.device, use_external_gpu_allocator: bool = True,
                       is_rocm_pytorch: bool = False, loglevel: int = 2):
        """Creates and returns the session configuration to be used for the ExecutionAgent"""

        providers = None
        provider_options = None
        if device.type == 'cuda':
            # Configure the InferenceSessions to use the specific GPU on which the model is placed.
            providers = (["ROCMExecutionProvider"] if is_rocm_pytorch else ["CUDAExecutionProvider"])
            providers.append("CPUExecutionProvider")
            if use_external_gpu_allocator and torch.cuda.is_available():
                # CPP extension to get torch GPU allocator's alloc and free function addresses
                from onnxruntime.training.ortmodule.torch_cpp_extensions import torch_gpu_allocator
                torch_alloc = torch_gpu_allocator.gpu_caching_allocator_raw_alloc_address()
                torch_free = torch_gpu_allocator.gpu_caching_allocator_raw_delete_address()
                provider_options = [{"device_id": str(device.index),
                                    "gpu_external_alloc": str(torch_alloc),
                                    "gpu_external_free": str(torch_free)}, {}]
            else:
                provider_options = [{"device_id": str(device.index)}, {}]
        elif device.type == 'cpu':
            providers = ["CPUExecutionProvider"]
            provider_options = [{}]

        session_options = onnxruntime.SessionOptions()
        session_options.enable_mem_pattern = False
        session_options.enable_mem_reuse = False
        session_options.use_deterministic_compute = False
        # default to PRIORITY_BASED execution order
        session_options.execution_order = onnxruntime.ExecutionOrder.PRIORITY_BASED
        # 0:Verbose, 1:Info, 2:Warning. 3:Error, 4:Fatal. Default is 2.
        session_options.log_severity_level = int(loglevel)

        self.session_options = session_options
        self.providers = providers
        self.provider_options = provider_options
```

How about cutting the extra `get_session_config` and just do everything o SessionConfig init?

Reply:
Makes sense. By the way, do you need `object`? Are we using python2 anywhere?

Reply:
Can we make it a dataclass?

Reply:
Python 2 is no longer supported. I think dataclass is fine